### PR TITLE
feat:Control for Changing Registered Email Address For An Account

### PR DESCRIPTION
**Pull request by philCryoport**
Hi all,

I originally submitted these proposed instructions for changing a system user's registered email address [to the Cheat Sheets repo](https://github.com/OWASP/CheatSheetSeries/pull/1245) -- and was told that the content was good, but it was too verbose for a cheat sheet (I concur in hindsight).

I think members of the technical community DO need an "argument from authority" describing in detail this process.

There are posts in StackExchange, but StackExchange is not an authority.  

I searched NIST -- an authority -- no luck.  Same for CISA.

OWASP is such an authority -- and I hope all y'all agree with the process I propose -- and accept this pull request.

Thank you for your consideration!

/Phil Schlesinger
Lifetime Member -- joined via the Orange County, CA OWASP meetup

### CODE REVIEW

**Comment by philCryoport** [comment]
> Thanks. It's gonna take me a bit to get a good window to review this.

No worries @kingthorin.  Just so I don't follow-up too early, by when _should_ I follow-up if I haven't had a review?

**Comment by kingthorin** [comment]
Ummm hit me up on Wednesday if you haven't heard.

**Review by kingthorin — COMMENTED** [review]
Could you replace the walls of text with flow diagrams?

File: pages/controls/Changing_Registered_Email_Address_For_An_Account.md [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+{% include writers.html %}
+
+People's email addresses regularly change.  The process below is the recommended method to implement in a system to handle that situation.  
+
+NOTE: The process is less stringent when using [Multifactor Authentication](https://cheatsheetseries.owasp.org/cheatsheets/Multifactor_Authentication_Cheat_Sheet.html) because the proof-of-identity is stronger than just using a password.
+
+## Process (Depending on a System's Level of Authentication)
+
+### Recommended Process If The User HAS [Multifactor Authentication](https://cheatsheetseries.owasp.org/cheatsheets/Multifactor_Authentication_Cheat_Sheet.html) Enabled
```
Comment (kingthorin) [inline pages/controls/Changing_Registered_Email_Address_For_An_Account.md:18]: I'd suggest not putting links in headings.

**Review by kingthorin — COMMENTED** [review]
The text/processes seem okay to me.

File: pages/controls/Changing_Registered_Email_Address_For_An_Account.md (hunk 1 above)
Comment (philCryoport) [reply pages/controls/Changing_Registered_Email_Address_For_An_Account.md]: Done

**Comment by philCryoport** [comment]
> Could you replace the walls of text with flow diagrams?

Hi @kingthorin do you mean you want me to replace the step-by-step texts with flowcharts?

**Comment by kingthorin** [comment]
That or at least add flow charts. The giant chunks of text are hard to consume.

File: pages/controls/Changing_Registered_Email_Address_For_An_Account.md (hunk 1 above)
Comment (kingthorin) [reply pages/controls/Changing_Registered_Email_Address_For_An_Account.md]: If you feel the link is important then just add a lead in sentence before the process

**Comment by kingthorin** [comment]
Will that actually render with GitHub pages?

https://owasp.org/migration/ has info on setting up a local testing environment.

**Comment by mtesauro** [comment]
@philCryoport  FWIW, I wrote this a while ago to make locally testing easier: https://github.com/mtesauro/owasp-test-site

YMMV but I've found it useful for pages/edits I've done.

File: pages/controls/Changing_Registered_Email_Address_For_An_Account.md (hunk 1 above)
Comment (philCryoport) [reply pages/controls/Changing_Registered_Email_Address_For_An_Account.md]: Nah, it's ok.  I link to it later on in the paragraphs

**Comment by philCryoport** [comment]
> Will that actually render with GitHub pages?
> 
> [owasp.org/migration](https://owasp.org/migration/) has info on setting up a local testing environment.

My theory is:
* Write the PlantUML
* Use a viewer to turn it into a diagram
* Put the diagram in the `assets` folder
* When coding the image into the document, add non-renderable comments on how to modify the diagram

Ok by you?

**Comment by kingthorin** [comment]
Cool, sounds good to me.

**Comment by philCryoport** [comment]
Hi @kingthorin please review when you get a moment?

The flowchart isn't a panacea, but it does illustrate the steps required in a less-verbose way...

**Comment by kingthorin** [comment]
Thanks for working on this, I'm still not ignoring it, hope to get to it in the next few days.

**Comment by philCryoport** [comment]
> Thanks for working on this, I'm still not ignoring it, hope to get to it in the next few days.

Hi @kingthorin -- apologies if I'm following up too quickly...

**Comment by kingthorin** [comment]
All good, thanks for the reminder. I'll try to prioritize it for tomorrow.

**Comment by kingthorin** [comment]
Hey @philCryoport, I haven't been able to tackle this yet. It isn't forgotten but the weekend and holidays are likely going to keep me busy. I'll try to get back to it asap.

**Comment by philCryoport** [comment]
> Hey @philCryoport, I haven't been able to tackle this yet. It isn't forgotten but the weekend and holidays are likely going to keep me busy. I'll try to get back to it asap.

Happy Holidays @kingthorin !  If you can't get to it before New Years, the world will still be ok.

File: pages/controls/Changing_Registered_Email_Address_For_An_Account.md [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+IF a User of Google only has their account protected with a password, when someone submits a change of registered email address, if the submitter knows the User's password, [Google just sends a **notification-only** email message to the currently-registered email address](https://support.google.com/accounts/answer/55393?hl=en) for that Google User with a link for "I didn't request this email address change".
+
+This is risky:
+
+The User's password might be easily-guessable -- or might be a repeated password stolen from another website where the User is registered.
+
+Nevertheless, Google apparently is still ok with the risk that the request might be unauthorized.  Hopefully the real User will see the email message to the soon-to-be originally-registered email address and click the "this is unexpected" link before the "this is unexpected" nonce expires.
+
+- Regular Social Engineering Training Is Essential
```
Comment (kingthorin) [inline pages/controls/Changing_Registered_Email_Address_For_An_Account.md:152]:
```suggestion
### Regular Social Engineering Training Is Essential
```

File: pages/controls/Changing_Registered_Email_Address_For_An_Account.md [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+      - IF the nonce in that link has expired
+      - THEN the System shall:
+        - FIRST remove that now-detected-as-expired nonce from the System data-store.
+        - AND THEN state to the User that the request is invalid.
+   - In the case that the nonce in the link cannot be found:
+      - The System shall state to the User that the request is invalid.
+
+## Notes on the Above Processes
+
+- Google thinks differently
```
Comment (kingthorin) [inline pages/controls/Changing_Registered_Email_Address_For_An_Account.md:142]:
```suggestion
### Google thinks differently
```

File: assets/plantuml/controls/Changing_Registered_Email_Address_For_An_Account.puml
Comment (kingthorin) [inline assets/plantuml/controls/Changing_Registered_Email_Address_For_An_Account.puml:1]: Maybe move this within assets/images/controls? Add a src directory and add a README.md that just says something like "Contains the source UML for some flowcharts etc. In the can of plantuml the UML was rendered in another tool and screenshotted for use on the site".

`assets/images/controls/src/plantuml/<this file>`
`assets/images/controls/src/README.md`

**Review by kingthorin — COMMENTED** [review]
LGTM overall, just a few minor things.

File: pages/controls/Changing_Registered_Email_Address_For_An_Account.md (hunk 2 above)
Comment (philCryoport) [reply pages/controls/Changing_Registered_Email_Address_For_An_Account.md]: DONE

File: pages/controls/Changing_Registered_Email_Address_For_An_Account.md (hunk 3 above)
Comment (philCryoport) [reply pages/controls/Changing_Registered_Email_Address_For_An_Account.md]: DONE

File: assets/plantuml/controls/Changing_Registered_Email_Address_For_An_Account.puml
Comment (philCryoport) [reply assets/plantuml/controls/Changing_Registered_Email_Address_For_An_Account.puml:1]: Hi @kingthorin ok I think I took care of everything.  Please take a look -- and if it looks ok, please approve?  Thanks!

**Comment by philCryoport** [comment]
Hi @kingthorin ok I think I took care of everything.  Please take a look -- and if it looks ok, please approve?  Thanks!

**Comment by kingthorin** [comment]
Thanks, I'll try to get this knocked of shortly.

**Review by kingthorin — COMMENTED** [review]
I'm good with this. I've asked on OWASP slack for someone else to have a look and provide their opinion so that I'm not unilaterally deciding to be an authority on such things :wink:

**Comment by avidouglen** [comment]
So I took a look at this. Each part of this makes sense, and overall it is a cohesive process. 👍

But it does feel overcomplicated, with unpleasant UX and lots of pieces that are easy to get wrong. 
Also, it doesn't account for the original email being lost or compromised. Which might be fine, but then let's make that explicit - otherwise it does feel like it is trying to be fully comprehensive :-) 

The other thing that bothers me about this - and it absolutely depends on context, what type of app, etc - this flow seems to assume/expect there to be only a single, authoritative email address for the user. Not just identification, but the identity itself. Contrast that to apps (e.g. like Github) wherein you can add multiple email addresses to a single user account.  It is an architectural consideration to be sure, but that changes a lot of perspective as well. 
E.g. just adding a 2nd email address to an existing account, you would simply reauthenticate the user, and then send notification-only email to all the other existing addresses on the account (as you do note that Google does), with an option to react or cancel (eg something like "If this wasnt you, let us know"). 
And cannot remove an address or change the primary, until the 2nd is approved. 

I would like to make a suggestion, to review other large / reputable / reliable sites to compare this proposed flow with what they actually do, and their existing threat model for that flow. E.g. you mentioned Google, I mentioned Github - worth digging in a bit. My assertion is that we'll find that very few actually go through all this. 
That said, to repeat my earlier comment: it really depends on context and type of app :-)

**Comment by philCryoport** [comment]
Hi @avidouglen -- thank you for commenting.

> So I took a look at this. Each part of this makes sense, and overall it is a cohesive process. 👍

Thank you!

 
> But it does feel overcomplicated, with unpleasant UX and lots of pieces that are easy to get wrong. 

Re: "lots of pieces that are easy to get wrong" -- just to elucidate the reality:

Most public websites DON'T have MFA -- and of those that do, MFA is rarely required.

Thus, in nearly every case -- with the only thing stopping an attacker from not just taking over an account, but additionally hiding activities from the actual User -- the web version of a middle-of-the-night SIM swap attack -- is the actual User's control of their registered email address inbox.


Re: "unpleasant UX": the UX these pages that I'm describing would be just as pleasant or as unpleasant as the rest of the same system.  Wouldn't make a difference.


> Also, it doesn't account for the original email being lost or compromised. Which might be fine, but then let's make that explicit - otherwise it does feel like it is trying to be fully comprehensive :-)

What-If statements added.
 

> The other thing that bothers me about this - and it absolutely depends on context, what type of app, etc - this flow seems to assume/expect there to be only a single, authoritative email address for the user. Not just identification, but the identity itself. Contrast that to apps (e.g. like Github) wherein you can add multiple email addresses to a single user account. It is an architectural consideration to be sure, but that changes a lot of perspective as well. E.g. just adding a 2nd email address to an existing account, you would simply reauthenticate the user, and then send notification-only email to all the other existing addresses on the account (as you do note that Google does), with an option to react or cancel (eg something like "If this wasnt you, let us know"). And cannot remove an address or change the primary, until the 2nd is approved.
>
> I would like to make a suggestion, to review other large / reputable / reliable sites to compare this proposed flow with what they actually do, and their existing threat model for that flow. E.g. you mentioned Google, I mentioned Github - worth digging in a bit. My assertion is that we'll find that very few actually go through all this. That said, to repeat my earlier comment: it really depends on context and type of app :-)


Forgive me, I'd like to get this magnum opus posted as it'll fit probably 99% of the sites that are out there -- including my current corporate need.

I've added an entry at the bottom of the page covering your above `1 User Account :: 1-* Registered Email Addresses` scenario, stating that this will be handled in a forthcoming version and encouraging content proposals.

**Comment by philCryoport** [comment]
@kingthorin if you're ok with the current content, please do approve it?

**Comment by avidouglen** [comment]
@philCryoport fair response! It doesn't need to be 100% optimized for everyone straight form the first version.  

One thing I would like to ask, is to be explicit about the scenario you are covering, straight at the beginning? Because if you are going for the "argue from authority", we should be advising for proper architectural structure, including 1-to-n accounts to email addresses (in most typical cases, this would be the right answer).   
Of course there are situations where that would not work, or more likely too hard to change that right now - and then that's where your scenario fits in, but it is not the ideal solution. (Yes I am all about the judicious compromise!) 
I.e. "If you only allow a single email address per account and this represents the user identity, then here is what you need to do: [the whole flow]". 

Do you think this is a fair caveat to add?

**Comment by avidouglen** [comment]
Btw re UX, these would of course be the same UI and design as the rest of the app, I was talking about the flows from the perspective of the user. 

And I would still be concerned about all this nonce juggling, irrelevant of MFA - but that is down to implementation, obviously. 

(Neither of these points are blockers IMO, but worth consideration for future).

**Comment by philCryoport** [comment]
> @philCryoport fair response! It doesn't need to be 100% optimized for everyone straight form the first version.
> 
> One thing I would like to ask, is to be explicit about the scenario you are covering, straight at the beginning? Because if you are going for the "argue from authority", we should be advising for proper architectural structure, including 1-to-n accounts to email addresses (in most typical cases, this would be the right answer). Of course there are situations where that would not work, or more likely too hard to change that right now - and then that's where your scenario fits in, but it is not the ideal solution. (Yes I am all about the pragmatic compromise!) I.e. "If you only allow a single email address per account and this represents the user identity, then here is what you need to do: [the whole flow]".
> 
> Do you think this is a fair caveat to add?

Sure @avidouglen . I've added the clarity to the section headings -- that way when we add the multiple-registered email addresses scenarios, the section headings are the differentiator.  Ok by you?

image
-- and --
image

**Comment by avidouglen** [comment]
Great, thanks! I like it!

**Review by kingthorin — APPROVED** [review]

**Pull request merged**